### ISSUE TITLE
WatchMate High background CPU when not connected.

### ISSUE THREAD

**Issue by FuzzyExpress** [azymohliad/watchmate#57]
My PineTime is not connected to my laptop, and WatchMate is using 100% of a core. This is after having had restarted the BlueTooth driver several times.
![Screenshot_20240125_202701](https://github.com/azymohliad/watchmate/assets/127012288/0a8b7452-4141-43da-b924-2f2f6488bf8f)
![image](https://github.com/azymohliad/watchmate/assets/127012288/2805202f-e250-4294-8e6f-51883eb1dc23)

I am running Kubuntu 23.10 | KDE Plasma 5.27.8 | Wayland | 6.6.8-x64v3-xanmod1 Kernel.
Let me know how to provide more information and I'll be sure to add it.

**Comment by azymohliad** [comment]
Thanks for the report. I noticed it too after leaving it in the background for a long time. No ideas yet on what's causing it. Will try to take a closer look in Feb

**Comment by FuzzyExpress** [comment]
I have looked at is several times today, no peaking today, connected or not. I never seems to reconnect to my watch after sleeping, but that's not really a problem, though it does have higher CPU use while not connected, when it was connected it would occasionally show up on `top` with 0.3%, when unconnected it makes a consistent 1.3% use. This comment is collected while using Xorg, but I don't think the display server affects this app in this scenario due to it not using the GUI.

**Comment by FuzzyExpress** [comment]
![Screenshot_20240130_004639](https://github.com/azymohliad/watchmate/assets/127012288/4f184865-3545-4d91-b47e-10eb9202a2a3)
Does WatchMate use any GPU acceleration perhaps (or a bug that would cause it to try to use the dGPU)? It shares the same symptoms as other dGPU apps that try to use my Nvidia card after sleeping (which breaks the GPU driver)

**Comment by azymohliad** [comment]
Hm.. That's an interesting theory to check. GTK4 can use it. Are you using Flatpak or a native distro package? For flatpak you can try to disable GPU acceleration permission with Flatseal or  `flatpak override --nodevice=dri io.gitlab.azymohliad.WatchMate`. Though I would expect other GTK4 apps to behave similarly if that's the problem...

**Comment by FuzzyExpress** [comment]
I am using Flatpak, I'll try disabling GPU acceleration and see if it has a effect on it. I haven't seen the other GTK app I use, EasyEffects, have a problem like this though. EasyEffects isn't installed via Flatpak, I wonder if Flatpak could have some reason to do with this. I'll be sure to let you know if the override fixes it or not.

**Comment by FuzzyExpress** [comment]
I believe I have confirmed that that is in fact what is happening. I applied the suggested flatpak override, but didn't reboot the app or my computer, and after sleeping it did the same thing again. After a rebooting and then sleeping it does not have this issue anymore it seems. Annoying that sleeping breaks everything wanting to use the dGPU, but glad to have this app fixed.

**Comment by azymohliad** [comment]
Nice, thanks a lot for figuring that out! I'll try to confirm this on my end too.

One potentially relevant difference between Watchmate and some other GTK4 apps is how the background mode is implemented: Watchmate simply hides the window, and some other apps I examined properly de-initialize the UI. The bad news is that Relm4 (the layer on top of GTK4 that Watchmate is built with) doesn't allow to implement background mode properly. So if that's the root cause, a proper fix might require a full rewrite in plain GTK4. I was considering it before, as it would bring some other benefits, and it shouldn't bee too hard, but I'm really not sure when I'll find the time and motivation.

**Comment by FuzzyExpress** [comment]
I just woke my laptop back up from sleep after having dinner and WatchMate is once again in the broken state. However this time the Nvidia stuff isn't broken; `nvidia-smi` `nvtop` and Blender Cycles work as intended.
![image](https://github.com/azymohliad/watchmate/assets/127012288/33d0f6ed-a9da-4617-82d3-f668dc56fe1d)
How bizarre.

**Comment by azymohliad** [comment]
Hi! I am now reasonably sure I found the root cause and fixed it in #62. After bluetooth adapter restart (e.g. after system suspend) and with the right timings, the device discovery could stuck in an infinite loop without yielding the thread.

I want to include some more fixes and additions into the next release and test it more, so I'm not sure when I'm gonna release it. But if you'd like to try this earlier, you can install [this test build from Flathub CI](https://github.com/flathub/io.gitlab.azymohliad.WatchMate/pull/12):

```
flatpak install https://dl.flathub.org/build-repo/87429/io.gitlab.azymohliad.WatchMate.flatpakref
```

**Comment by FuzzyExpress** [comment]
# OHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHH
That makes a **LOT** of  sense, I had a issue where my Bluetooth mouse wouldn't connect so I made an alias `rsbt` to restart the Bluetooth driver, I would run this many times a day, almost as much as I run the `blender` command.
The only thing I need now is to find the gosh darn charger, its been lost for prob 2 months now. If I ever find it or get a new one I'll be sure to test it on my SteamDeck (the laptop died).

**Comment by azymohliad** [comment]
Don't stress about testing it if it takes an extra effort. Of course it would be good to know if it helped, but there's no rush with that. This fix is useful anyway, and will be included in the next release. And if it turns out that it doesn't actually solve this issue, we'll just keep looking for the cause.

**Comment by FuzzyExpress** [comment]
No, you haven't caused me any stress. :) That was just my unnecessarily dramatic expression of my surprise that over the entire time we spent trying to figure it out, that I never once stopped to considered that it maybe wasn't a GUI problem but a Bluetooth problem. That makes sense given that other app's GPU Hanging problem has the same 100% CPU use 
that this Bluetooth loop problem had. There is a term for this, like narrow minded or something I don't remember exactly. 

Either way you also reminded me I had this watch, and that I should probably use it instead of it just being a whist weight? (I don't think have a paper weight for your wrist is exactly normal...)
Now that we know the problem (we think) the watch will be useful again.

This along side [#59 Auto Reconnect](https://github.com/azymohliad/watchmate/issues/59) putting this on multiple computers might be practical. I'll try this as well once I get my watch up and running.

Thanks for making this nice easy to use app.

**Comment by azymohliad** [comment]
FYI, I've released this fix as v0.5.2 (should appear on flathub in a few hours). It turns out all the other improvements I wanted to include in the next release will take more time than I expected.

**Comment by FuzzyExpress** [comment]
Oh, perfect timing! I just found my PineTime charger trying to dig my phone out of my bag after an exam.
I'll be sure to test it once it's available.

**Comment by FuzzyExpress** [comment]
Update: I have not yet seen the CPU spike issue happen at all with the new package! I've put my KDE Neon Unstable Mac and Steam Deck to sleep and resumed several times, no CPU spikes, well not from WatchMate anyways, which is great. (steam did it once). the watch still often does not reconnect after walking out of range or sleeping, but I can open WatchMate and reconnect it. Now I will note that I have not used the actual `sudo systemctl restart bluetooth` command, I've been using my mouse wired.

**Comment by azymohliad** [comment]
Ok, thanks, it's good to know that auto-reconnection still has issues (re #59). I recently learnt that there might be a more reliable and efficient way to auto-reconnect if the device is bonded. So the plan is to implement secure connection first (#56), and then try to improve auto-reconnection. I'm half-way through secure connection implementation, but a bit busy this weekend, so will probably get back to it not earlier than the next weekend.
